src/coreclr/jit/valuenum.h

```diff
@@ -594,6 +595,20 @@ class ValueNumStore
     //
     template <typename TArgVisitor>
     VNVisit VNVisitReachingVNs(ValueNum vn, TArgVisitor argVisitor)
+    {
+        // Fast-path: in most cases vn is not a phi definition
+        if (!IsPhiDef(vn))
+        {
+            return argVisitor(vn);
+        }
+        return VNVisitReachingVNsWorker(vn, argVisitor);
+    }
+
+private:
+
+    // Helper function for VNVisitReachingVNs
+    template <typename TArgVisitor>
+    VNVisit VNVisitReachingVNsWorker(ValueNum vn, TArgVisitor argVisitor)
```

Comment:
Here I split `VNVisitReachingVNs` into two functions. In most cases the input is not a phi so we run `argVisitor` just once. So we don't need to allocate an array (on stack)